Rebase on rayon-1.2.0 and rayon-core-1.6.0

cuviper opened the pull request:
This can't be directly merged, since I rebased, but I hope it's useful anyway.

r? @Zoxc

cuviper commented:
I think you could drop your main handler and scoped pools in favor of `build_scoped`, but for now I just carried that patch as-is.

https://docs.rs/rayon/1.1.0/rayon/struct.ThreadPoolBuilder.html#method.build_scoped

RalfJung commented on Cargo.toml:
```diff
@@ -1,26 +1,25 @@
-rayon-core = { version = "1.4", path = "rayon-core" }
-crossbeam-deque = "0.2.0"
+rustc-rayon-core = { version = "0.2", path = "rayon-core" }
+crossbeam-deque = "0.6.3"
```
It would be *awesome* to get this change landed as it would remove the dependency of librustc (and likely other crates) on old unsound versions of `memoffset`.

cuviper commented:
I've now rebased this on rayon 1.2.

I also dropped the `main_handler` and `scoped_pool` patches in favor of `build_scoped`. See the draft PR https://github.com/rust-lang/rust/pull/64358 -- the code doesn't actually change much for that.

cuviper commented:
The comparison to rayon-rs master may be more useful:
https://github.com/rayon-rs/rayon/compare/master...cuviper:rustc

cuviper commented:
@Zoxc, can you review this? It would be nice to land in rustc, if only for the dependency updates.

mati865 commented:
It still needs new crates.io release to make it's way to Rust.

cuviper commented:
@mati865 I think this did make it into rustc-rayon 0.3.0 -- in fact it was the reason for that release.

cuviper commented:
That's on the `crate-0.3.0` branch -- it looks like the `rustc` branch has since rebased onto rayon 1.3.0, and _that_ hasn't been published yet.
